### PR TITLE
🌱 Add testctl util

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: hack/tools/testctl/core/plugins.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		return nil, err
+	}
+
+	if config.Path == "" {
+		return nil, pkgerrors.New("path must be set when calling the importer plugin")
+	}
+
+	path, err := filepath.Abs(config.Path)
+	if err != nil {
+		return nil, pkgerrors.Errorf("failed to convert %s to an absolute path", config.Path)
```
Comment (sbueringer) [inline hack/tools/testctl/core/plugins.go]:
```suggestion
		return nil, pkgerrors.Wrapf(err, "failed to convert %s to an absolute path", config.Path)
```

File: hack/tools/testctl/core/runner.go [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+	Concurrency *int32 `json:"concurrency,omitempty"`
+
+	// failFast defines the behavior in case a test action fails.
+	// If not set, it defaults to false.
+	// Note that if concurrency is greater than 1, the system will try to complete other in-flight tests before exiting.
+	// TODO: think about using this for failures that could happen to a plugin in the middle of a list of plugins (run field). Currently it is always fail fast
+	FailFast *bool `json:"failFast,omitempty"`
+
+	// dryRun instructs plugins to run in dry run mode (no change applied).
+	// Note: SelectorPlugins are expected to return TestObjects no matter of the value of this field.
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go:66]: Let's also document the default behavior if the field is not set, like above

(same for other fields below)

File: hack/tools/testctl/core/runner.go [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+// Run the tests defined in the config using the given plugins.
+func Run(ctx context.Context, c client.Client, config []byte, plugins map[string]any) error {
+	for k, plugin := range plugins {
+		if _, ok := plugin.(SelectorPlugin); ok {
+			continue
+		}
+		if _, ok := plugin.(ExecutorPlugin); ok {
+			continue
+		}
+		return pkgerrors.Errorf("plugin %s must implement one of SelectorPlugin, ExecutorPlugin", k)
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go]: I think this validation logic allows to implement both interfaces in one plugin. Is this intentional? If not, should we block it?

File: hack/tools/testctl/main.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	restConfig, err := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{
+		Context: *context,
+	}).ClientConfig()
+	if err != nil {
+		return pkgerrors.Wrapf(err, "failed to create rest config from %q", configLoadingRules.GetExplicitFile())
+	}
+	restConfig.UserAgent = "testctl"
+	restConfig.QPS = 20
+	restConfig.Burst = 100
```
Comment (sbueringer) [inline hack/tools/testctl/main.go]: Is that enough? (our controllers are running with 1k & 2k to avoid client-side rate-limiting in scale tests)

File: hack/tools/testctl/core/runner.go [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+		configs: make(map[string]any),
+	}
+}
+
+// parseConfig parses the config file passed to the runner as input.
+// Note: during this operation, config fragments are stored in runner.configs so they can be retrieved later at run stage.
+func (r *runner) parseConfig(ctx context.Context, path *field.Path, plugin any, callStack []string, rawConfig []byte) error {
+	// TODO: in future we might want to injecting env variables / use templating
+
+	var err error
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go]:
```suggestion
```
nit. Let's define this var locally where it is needed?

File: hack/tools/testctl/core/runner.go [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+	if validator, ok := plugin.(CallStackValidatorPlugin); ok {
+		if err := validator.ValidateCallStack(ctx, callStack); err != nil {
+			return pkgerrors.Wrapf(err, "failed to validate call stack for %s", path.String())
+		}
+	}
+
+	// If this is a configurable plugin, read the custom plugin config.
+	// Note: When invoking plugin's ParseConfig method, drop all the fields that are managed from the runner itself,
+	// so the plugin authors can use Unmarshal strict.
+	if configurable, ok := plugin.(ConfigurablePlugin); ok {
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go]: What about also failing here if the plugin is not configurable but has some config?

File: hack/tools/testctl/core/runner.go [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+	if _, ok := plugin.(*importerPlugin); ok {
+		return
+	}
+
+	if strings.HasSuffix(msgPrefix, " ") {
+		msgPrefix += " "
+	}
+	if strings.HasPrefix(msgSuffix, " ") {
+		msgSuffix = " " + msgSuffix
+	}
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go]: Why is this needed? (maybe let's add a godoc)

File: hack/tools/testctl/core/runner.go [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+			return nil, pkgerrors.Wrapf(err, "failed to marshal %s.runConfig", path)
+		}
+
+		if err := yaml.UnmarshalStrict(cfgByte, &runConfig); err != nil {
+			return nil, pkgerrors.Wrapf(err, "failed to unmarshal %s.runConfig", path)
+		}
+	}
+
+	// Use top level runConfig as a default when it makes sense.
+	if path.String() != rootPath.String() {
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go:487]: I don't follow why we are defaulting these fields, but not the other ones in the RunConfig struct

File: hack/tools/testctl/core/runner.go [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+	if generator, ok := plugin.(MessageGeneratorPlugin); ok {
+		if generatedMessage, err := generator.GenerateMessage(objects, pluginConfig); err == nil {
+			msg = generatedMessage
+		}
+	}
+	r.prompter(ctx, fmt.Sprintf("%s%s%s", msgPrefix, msg, msgSuffix))
+}
+
+// getRunConfig parses a runConfig field if defined, otherwise returns a default RunConfig.
+func (r *runner) getRunConfig(_ context.Context, path *field.Path, rawConfig map[string]any) (*RunConfig, error) {
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go:472]: For my understanding. This is reading the runConfig on the current level and is defaulting with the root one, but it does not inherit the values from in between, right?

File: hack/tools/testctl/README.md [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+
+A more complete example, scaling up a Cluster's control plane and workers, then rolling back
+on failure, then upgrading:
+
+```yaml
+runConfig:
+  concurrency: 2
+  dryRun: false
+
+run:
```
Comment (sbueringer) [inline hack/tools/testctl/README.md:293]: Maybe slightly extend to show how runConfig can be set inside the plugin hierarchy

File: hack/tools/testctl/core/runner.go [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+var rootPath = field.NewPath("$")
+
+// TestObjects is a map of key/objects to be passed to a plugin, e.g. "Cluster":..., "MachineDeployment":...
+// Note: We are using this approach instead of adding info to the context because it is easier to debug.
+type TestObjects map[string]any
+
+// RunConfig defines the run config to be used when invoking plugins.
+// Note: For the sake of simplicity there is a unique RunConfig for all the plugins; as a consequence some field in this struct
+// might not be relevant for a specific plugin. Some field in this struct are managed by the runner itself, e.g. concurrency.
+type RunConfig struct {
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner.go:48]: We probably don't want to change it as this point, but I assume the code in this file would be simpler if the config file would be ~ mostly typed.

Let's make sure the example YAML in README.md covers all relevant config fields/variants. That should help to make up for the lack of full config structs that could be used as reference when writing a config file

File: hack/tools/testctl/core/runner_test.go [hunk 12]
```diff
@@ -0,0 +1,869 @@
...
+			g.Expect(plugin).ToNot(BeNil())
+
+			rawConfig, err := yaml.Marshal(tt.config)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			// TODO: check error message
+			err = r.parseConfig(t.Context(), rootPath, plugin, nil, rawConfig)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			// TODO: check error message
```
Comment (sbueringer) [inline hack/tools/testctl/core/runner_test.go]: Seems implemented

File: hack/tools/testctl/plugins/cluster_delete.go [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+	cluster, err := p.unwrapTestObjects(objects)
+	if err != nil {
+		return err
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("Cluster", klog.KObj(cluster))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		log.Info("Upgrade Cluster action skipped, Cluster already is deleting")
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/cluster_delete.go]: Should this be "Delete cluster action skipped, ..."

File: hack/tools/testctl/plugins/cluster_delete.go [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+		return nil
+	}
+
+	log.Info("Deleting Cluster")
+	if ptr.Deref(runConfig.DryRun, false) {
+		return nil
+	}
+
+	if err := c.Delete(ctx, cluster); err != nil {
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/cluster_delete.go:63]: maybe handle NotFound errors and return directly if one occurs?

File: hack/tools/testctl/plugins/cluster_delete.go [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+		return pkgerrors.Wrapf(err, "failed to delete Cluster %s", klog.KObj(cluster))
+	}
+
+	if ptr.Deref(runConfig.SkipWait, false) {
+		return nil
+	}
+
+	log.Info("Waiting for Cluster to be deleted")
+	if err := wait.PollUntilContextTimeout(ctx, 5*time.Second, 5*time.Minute, false, func(ctx context.Context) (done bool, err error) {
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/cluster_delete.go]: should this use the timeout from the runConfig? (same in other files, e.g. cluster_upgrade.go)

File: hack/tools/testctl/plugins/cluster_selector.go [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+func (p *ClusterSelectorPlugin) unwrapConfig(pluginConfigUntyped any) *ClusterSelectorPluginConfig {
+	config := &ClusterSelectorPluginConfig{}
+	if pluginConfigUntyped != nil {
+		config = pluginConfigUntyped.(*ClusterSelectorPluginConfig)
+	}
+	return config
+}
+
+// UnwrapClusterTestObject extracts the Cluster from TestObjects.
+func UnwrapClusterTestObject(objects core.TestObjects) (*clusterv1.Cluster, error) {
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/cluster_selector.go:110]: Maybe let's move this to util.go? It's used in a lot of places but not in this file

File: hack/tools/testctl/plugins/cluster_upgrade.go [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+// ValidateCallStack validates the call stack for the plugin.
+func (p ClusterUpgradePlugin) ValidateCallStack(_ context.Context, callStack []string) error {
+	if !slices.Contains(callStack, clusterSelectorPluginKey) {
+		return pkgerrors.Errorf("this plugin can only be called as a child of %s", clusterSelectorPluginKey)
+	}
+	return nil
+}
+
+func (p ClusterUpgradePlugin) unwrapConfigAndTestObjects(objects core.TestObjects, pluginConfigUntyped any) (*ClusterUpgradePluginConfig, *clusterv1.Cluster, error) {
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/cluster_upgrade.go]: Up to you but I think instead of having duplicate methods we could probably have one generic func in util.go

File: hack/tools/testctl/plugins/cluster_delete.go [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		return err
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("Cluster", klog.KObj(cluster))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		log.Info("Upgrade Cluster action skipped, Cluster already is deleting")
+		return nil
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/cluster_delete.go]: Should this still wait if skipWait is false? (similar in other cases like cluster_upgrade.go)

I think controlplane_scale.go is already implemented this way.

Would be good to have this consistent one way or the other

File: hack/tools/testctl/plugins/controlplane_scale.go [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+	log.Info(fmt.Sprintf("Scaling KubeadmControlPlane from %d to %d replicas", currentReplicas, replicas))
+	if ptr.Deref(runConfig.DryRun, false) {
+		return nil
+	}
+
+	original := cluster.DeepCopy()
+	cluster.Spec.Topology.ControlPlane.Replicas = ptr.To(replicas)
+	if err := c.Patch(ctx, cluster, client.MergeFrom(original)); err != nil {
+		return pkgerrors.Wrapf(err, "failed to patch Cluster %s", klog.KObj(cluster))
+	}
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/controlplane_scale.go]: Should this Patch be skipped if replicas is already at the correct value?

File: hack/tools/testctl/plugins/controlplane_selector.go [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api-provider-vsphere/hack/tools/testctl/core"
+)
+
+const (
+	controlPlaneSelectorPluginKey = "select.controlPlane.testctl"
+
+	// controlPlaneObjectKey is the key used to identify the ControlPlane object in TestObjects.
+	controlPlaneObjectKey = "ControlPlane"
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/controlplane_selector.go]: Probably export this as well as ClusterObjectKey is already exported? (or unexport both)

(same for machineDeploymentSelectorObjectKey)

File: hack/tools/testctl/plugins/machinedeployment_create.go [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+	Template clusterv1.MachineDeploymentTopology `json:"template,omitempty"`
+}
+
+func init() {
+	registerPlugin(machineDeploymentCreatePluginKey, &MachineDeploymentCreatePlugin{})
+}
+
+// Exec executes the plugin for the given TestObjects.
+func (p MachineDeploymentCreatePlugin) Exec(ctx context.Context, c client.Client, objects core.TestObjects, pluginConfigUntyped any, runConfig core.RunConfig) error {
+	config, cluster, err := p.getInfo(objects, pluginConfigUntyped)
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/machinedeployment_create.go]: Let's rename getInfo to unwrapConfigAndTestObjects. If I see correctly in other plugins it is called that

File: hack/tools/testctl/plugins/machinedeployment_delete.go [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+// Exec executes the plugin for the given TestObjects.
+func (p MachineDeploymentDeletePlugin) Exec(ctx context.Context, c client.Client, objects core.TestObjects, _ any, runConfig core.RunConfig) error {
+	cluster, machineDeployment, mdTopologyIndex, err := p.unwrapTestObjects(objects)
+	if err != nil {
+		return err
+	}
+
+	if mdTopologyIndex == -1 {
+		// TODO: Log? MD CR still exists, but it has been already removed from the cluster.
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/machinedeployment_delete.go]: If the MD still exists we should maybe also wait for it to go away

File: hack/tools/testctl/plugins/machinedeployment_delete.go [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+var _ core.ExecutorPlugin = &MachineDeploymentDeletePlugin{}
+var _ core.MessageGeneratorPlugin = &MachineDeploymentDeletePlugin{}
+var _ core.CallStackValidatorPlugin = &MachineDeploymentDeletePlugin{}
+
+func init() {
+	registerPlugin(machineDeploymentDeletePluginKey, &MachineDeploymentDeletePlugin{})
+}
+
+// Exec executes the plugin for the given TestObjects.
+func (p MachineDeploymentDeletePlugin) Exec(ctx context.Context, c client.Client, objects core.TestObjects, _ any, runConfig core.RunConfig) error {
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/machinedeployment_delete.go:52]: Is this being run in parallel for multiple MDs of the same Cluster? If yes, is this concurrency safe? (specifically the part where we delete MD by index)

Are we getting conflict errors? If yes, how do we handle them? (same for the MD scale plugin)

File: hack/tools/testctl/plugins/machinedeployment_delete.go [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+	return nil
+}
+
+// GenerateMessage generates a message text for the plugin call.
+func (p MachineDeploymentDeletePlugin) GenerateMessage(objects core.TestObjects, _ any) (string, error) {
+	cluster, machineDeployment, _, err := p.unwrapTestObjects(objects)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("Delete MachineDeployments %s in Cluster %s", klog.KObj(machineDeployment), klog.KObj(cluster)), nil
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/machinedeployment_delete.go]:
```suggestion
	return fmt.Sprintf("Delete MachineDeployment %s in Cluster %s", klog.KObj(machineDeployment), klog.KObj(cluster)), nil
```

File: hack/tools/testctl/plugins/machinedeployment_scale.go [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+	log.Info(fmt.Sprintf("Scaling MachineDeployment from %d to %d replicas", currentReplicas, replicas))
+	if ptr.Deref(runConfig.DryRun, false) {
+		return nil
+	}
+
+	original := cluster.DeepCopy()
+	cluster.Spec.Topology.Workers.MachineDeployments[mdTopologyIndex].Replicas = ptr.To(replicas)
+	if err := c.Patch(ctx, cluster, client.MergeFrom(original)); err != nil {
+		return pkgerrors.Wrapf(err, "failed to patch Cluster %s", klog.KObj(cluster))
+	}
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/machinedeployment_scale.go]: Should we skip the patch if the replicas field is already set to the correct value?

File: hack/tools/testctl/plugins/machinedeployment_scale.go [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+
+// GenerateMessage generates a message text for the plugin call.
+func (p MachineDeploymentScalePlugin) GenerateMessage(objects core.TestObjects, pluginConfigUntyped any) (string, error) {
+	config, cluster, machineDeployment, mdTopologyIndex, err := p.unwrapConfigAndTestObjects(objects, pluginConfigUntyped)
+	if err != nil {
+		return "", err
+	}
+
+	replicas := targetMachineDeploymentReplicas(cluster, mdTopologyIndex, config)
+	return fmt.Sprintf("Scale MachineDeployments %s to %d replicas in Cluster %s", klog.KObj(machineDeployment), replicas, klog.KObj(cluster)), nil
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/machinedeployment_scale.go]:
```suggestion
	return fmt.Sprintf("Scale MachineDeployment %s to %d replicas in Cluster %s", klog.KObj(machineDeployment), replicas, klog.KObj(cluster)), nil
```

File: hack/tools/testctl/plugins/util.go [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+			continue
+		}
+
+		clusters = append(clusters, &cluster)
+		if limit != nil && len(clusters) >= int(ptr.Deref(limit, 0)) {
+			break
+		}
+	}
+
+	sort.Slice(clusters, func(i, j int) bool {
```
Comment (sbueringer) [inline hack/tools/testctl/plugins/util.go:57]: I wonder if the sort should happen before we filter out in l.50-57

Same in getMachineDeployments

File: hack/tools/testctl/core/runner.go (hunk 3 above)
Comment (fabriziopandini) [reply hack/tools/testctl/core/runner.go]: Fair point.
Implementing only one is simpler to reason about, so I have blocked it for this first iteration

File: hack/tools/testctl/main.go (hunk 4 above)
Comment (fabriziopandini) [reply hack/tools/testctl/main.go]: increased

File: hack/tools/testctl/core/runner.go (hunk 7 above)
Comment (fabriziopandini) [reply hack/tools/testctl/core/runner.go]: Cleaned up a bit, PTAL

File: hack/tools/testctl/core/runner.go (hunk 8 above)
Comment (fabriziopandini) [reply hack/tools/testctl/core/runner.go:487]: While using the tool, I found useful to be able to set some RunConfig flag globally, so I added them here (dry-run, debug, debug-on-error).

Setting other RunConfig flags globally seems to have less sense (limit, concurrency, failFast, timeout)

I have added a note to the field description to clarify ^^^, but open to suggestions

File: hack/tools/testctl/core/runner.go (hunk 9 above)
Comment (fabriziopandini) [reply hack/tools/testctl/core/runner.go:472]: that's correct.

Each plugin has it's own RunConfig, but I found also useful to be able to set some values globally (dry-run, debug, debug-on-error, intervals).

I think we should not make it more complex - aka introduce a fully hierarchical configuration model - in this first iteration.

File: hack/tools/testctl/core/runner.go (hunk 11 above)
Comment (fabriziopandini) [reply hack/tools/testctl/core/runner.go:48]: The config is intentionally extensible and only partially typed to support:
- adding plugins with their own custom configs
- combining plugins at will

The README has documentation for all the fields, I would avoid to make YAML files in README more complicated at this stage and instead have a long term goal to add a folder with concrete/reusable examples over time

File: hack/tools/testctl/plugins/cluster_selector.go (hunk 16 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/cluster_selector.go:110]: I think it makes more sense to have it here, near to where we set Cluster entry in the TestObjects

File: hack/tools/testctl/plugins/cluster_delete.go (hunk 18 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/cluster_delete.go]: That's a fair point, and I think we should always try to check if operation actually completed, and otherwise wait.

I have fixed for cluster delete, but if that's ok I would defer fixing for other actions because changes are more invasive and this PR is already big (I have added TODOs to keep track where changes are required)

File: hack/tools/testctl/plugins/controlplane_scale.go (hunk 19 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/controlplane_scale.go]: It is already skipping patch if we are at the target state (replicas set to the right value + current replicas aligned). But ok

File: hack/tools/testctl/plugins/machinedeployment_delete.go (hunk 22 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/machinedeployment_delete.go]: Tracked as a TODO (similar to other check if operation actually completed)

File: hack/tools/testctl/plugins/machinedeployment_delete.go (hunk 23 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/machinedeployment_delete.go:52]: That's a great point.

I think that a reasonable solution for a first iteration could be to try enforce that some plugins (e.g. MD delete) can only be run without parallelism. 

This requires to extend ValidateCallStack, so I have tracked as a TODO.

WRT to MD scale plugin, I don't think the same issue apply.

Note: there will some edge case like a configuration where it first delete an MD and then it tries to scale it up or down, but also I don't think it is worth the effort to try to make this test utility config bullet proof.

File: hack/tools/testctl/plugins/machinedeployment_scale.go (hunk 25 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/machinedeployment_scale.go]: It is already skipping patch if we are at the target state (replicas set to the right value + current replicas aligned). But ok

File: hack/tools/testctl/plugins/util.go (hunk 27 above)
Comment (fabriziopandini) [reply hack/tools/testctl/plugins/util.go:57]: Reviewed, PTAL

File: hack/tools/testctl/core/runner.go (hunk 7 above)
Comment (sbueringer) [reply hack/tools/testctl/core/runner.go]: Thx!

File: hack/tools/testctl/core/runner.go (hunk 8 above)
Comment (sbueringer) [reply hack/tools/testctl/core/runner.go:487]: Alrighty! Thx!

File: hack/tools/testctl/core/runner.go (hunk 9 above)
Comment (sbueringer) [reply hack/tools/testctl/core/runner.go:472]: Got it, that's fine.

File: hack/tools/testctl/core/runner.go (hunk 11 above)
Comment (sbueringer) [reply hack/tools/testctl/core/runner.go:48]: Alright, that's okay

File: hack/tools/testctl/plugins/cluster_selector.go (hunk 16 above)
Comment (sbueringer) [reply hack/tools/testctl/plugins/cluster_selector.go:110]: Got it, makes sense

File: hack/tools/testctl/plugins/cluster_delete.go (hunk 18 above)
Comment (sbueringer) [reply hack/tools/testctl/plugins/cluster_delete.go]: Absolutely, that's fine

File: hack/tools/testctl/plugins/controlplane_scale.go (hunk 19 above)
Comment (sbueringer) [reply hack/tools/testctl/plugins/controlplane_scale.go]: Yup. I was thinking about the case where replicas is set to the right value but currentReplicas != config.Replicas

File: hack/tools/testctl/plugins/machinedeployment_delete.go (hunk 23 above)
Comment (sbueringer) [reply hack/tools/testctl/plugins/machinedeployment_delete.go:52]: I assume this was addressed with "// TODO: fail if this plugin is run in parallel" (fine for me)

**Comment by sbueringer** [comment]
Thx, very nice!

/lgtm
/approve
